### PR TITLE
Fix `detect_hardhat_artifact_dir` in externalTests scripts

### PR DESCRIPTION
**Pull request by r0qs**
While working on https://github.com/ethereum/solidity/pull/14880, I noticed that the ENS external test has been failing silently for quite some time. 

The issue is from the fact that the `detect_hardhat_artifact_dir function` was returning an empty string, causing `bytecode_size_json_from_hardhat_artifacts` to use an incorrect path. This PR addresses that and also another issue within the same external test script, where it was attempting to run `sed` on a directory instead of a file.

See:  https://app.circleci.com/pipelines/github/ethereum/solidity/32974/workflows/f902a748-59ab-4eac-8367-8d71095744be/jobs/1480350?invite=true#step-110-61165_76
```
Skipping test function...
Cannot determine Hardhat artifact location. Both build/ and artifacts/ exist
jq: error: Could not open file /build-info/*.json: No such file or directory
Done.
```

And: https://app.circleci.com/pipelines/github/ethereum/solidity/32974/workflows/f902a748-59ab-4eac-8367-8d71095744be/jobs/1480350?invite=true#step-110-17512_102
```
sed: couldn't edit ./node_modules/@ensdomains/buffer/artifacts/hardhat/console.sol: not a regular file
sed: couldn't edit ./node_modules/@ensdomains/buffer/artifacts/contracts/Buffer.sol: not a regular file
sed: couldn't edit ./node_modules/@ensdomains/buffer/artifacts/contracts/Migrations.sol: not a regular file
sed: couldn't edit ./node_modules/@ensdomains/buffer/artifacts/test/TestBuffer.sol: not a regular file
sed: couldn't edit ./node_modules/@ensdomains/solsha1/artifacts/contracts/SHA1.sol: not a regular file
sed: couldn't edit ./node_modules/@ensdomains/solsha1/artifacts/test/mocks/SHA1Test.sol: not a regular file
```

### CODE REVIEW

**Comment by r0qs** [comment]
Hardhat and OpenZeppelin errors should be fixed by https://github.com/ethereum/solidity/pull/14875 and https://github.com/ethereum/solidity/pull/14880 respectively.

File: scripts/externalTests/common.sh [hunk 1]
```diff
@@ -572,15 +572,12 @@ function gas_report_to_json
 
 function detect_hardhat_artifact_dir
```
Comment (cameel) [inline scripts/externalTests/common.sh]:
```suggestion
function detect_hardhat_artifact_dirs
```

File: scripts/externalTests/common.sh [hunk 2]
```diff
@@ -605,16 +602,17 @@ function bytecode_size_json_from_truffle_artifacts
...
-        # Each artifact contains Standard JSON output under the `output` key.
-        # Process it into a dict of the form `{"<file>": {"<contract>": <size>}}`,
-        # Note that one Hardhat artifact often represents multiple input files.
-        jq '.output.contracts | to_entries[] | {
-            "\(.key)": .value | to_entries[] | {
-                "\(.key)": (.value.evm.bytecode.object | length / 2)
-            }
-        }' "$artifact"
+    for artifacts in $(detect_hardhat_artifact_dir); do
+        for artifact in "$artifacts"/build-info/*.json; do
```
Comment (cameel) [inline scripts/externalTests/common.sh]:
```suggestion
    for artifact_dir in $(detect_hardhat_artifact_dirs); do
        for artifact in "$artifact_dir"/build-info/*.json; do
```

File: scripts/externalTests/common.sh [hunk 3]
```diff
@@ -572,15 +572,12 @@ function gas_report_to_json
...
-        echo -n artifacts
-    else
-        fail "Hardhat build artifacts not found."
-    fi
+    # NOTE: The artifacts path is a configured parameter in Hardhat, so the below may fail for new external tests
+    # See: https://hardhat.org/hardhat-runner/docs/config#path-configuration
+    local artifact_dir=()
+    [[ -e build/artifacts ]] && artifact_dir+=("build/artifacts")
+    [[ -e artifacts/ ]] && artifact_dir+=("artifacts")
+    echo -n "${artifact_dir[@]}"
```
Comment (cameel) [inline scripts/externalTests/common.sh]: I'm not sure this this fixes things.

I see that the issue with the original code is that `fail` does not actually interrupt the script. Not sure why exactly. Maybe we should make `fail` do `exit` instead of `return`... The reason it does not do that now is that `exit` exits unconditionally, so it won't let you chain things that fail this way using `||`, which is sometimes handy and I think we rely on that in a few places. Maybe we should just define a variant like `hardFail` that does exit for those cases where the failure represents something closer to an assert than a recoverable error :)

But anyway, the current version differs only in that it never fails but instead may return one or both of these dirs. How does it fix ENS? Doesn't it just mean that now we now fail to detect its artifact dir and just do not inspect its artifacts?

File: test/externalTests/ens.sh [hunk 4]
```diff
@@ -81,7 +81,7 @@ function ens_test
     sed -i "s|it\(('Cannot be called if CANNOT_CREATE_SUBDOMAIN is burned and is a new subdomain',\)|it.skip\1|g" test/wrapper/NameWrapper.js
     sed -i "s|it\(('Cannot be called if PARENT_CANNOT_CONTROL is burned and is an existing subdomain',\)|it.skip\1|g" test/wrapper/NameWrapper.js
 
-    find . -name "*.sol" -exec sed -i -e 's/^\(\s*\)\(assembly\)/\1\/\/\/ @solidity memory-safe-assembly\n\1\2/' '{}' \;
+    find . -name "*.sol" -type f -exec sed -i -e 's/^\(\s*\)\(assembly\)/\1\/\/\/ @solidity memory-safe-assembly\n\1\2/' '{}' \;
```
Comment (cameel) [inline test/externalTests/ens.sh:84]: This is an independent fix, it should be in a separate commit with its own description.

File: scripts/externalTests/common.sh (hunk 3 above)
Comment (r0qs) [reply scripts/externalTests/common.sh]: > I'm not sure this this fixes things.
> 
> I see that the issue with the original code is that `fail` does not actually interrupt the script. Not sure why exactly. Maybe we should make `fail` do `exit` instead of `return`... The reason it does not do that now is that `exit` exits unconditionally, so it won't let you chain things that fail this way using `||`, which is sometimes handy and I think we rely on that in a few places. Maybe we should just define a variant like `hardFail` that does exit for those cases where the failure represents something closer to an assert than a recoverable error :)
>
> But anyway, the current version differs only in that it never fails but instead may return one or both of these dirs. How does it fix ENS? Doesn't it just mean that now we now fail to detect its artifact dir and just do not inspect its artifacts?

Because ENS has both artifact locations, namely `./build/artifacts` and `./artifacts/`, my change now utilizes both if they exist, and it does nothing if they don't. Therefore, we don't really need to hard fail in the first case, and for the latter, since no artifacts were found, we can just let the tests fail by itself.

However, it's possible that artifacts may be located elsewhere, and to locate them, we would need to search the project path for any artifacts directory and assume that they are generated by Hardhat, which may not always be accurate and could lead to other failures.

I believe it's better to take no action if we cannot find artifacts where we expect them and allow the test to fail if they exist in a different path and for some reason cause issues.

Comment (cameel) [reply scripts/externalTests/common.sh]: Ah, ok. I assumed that in case of ENS if was hitting the case with both dirs missing. This makes more sense.

In that case it sounds like returning two dirs indeed is the right fix.

> I believe it's better to take no action if we cannot find artifacts where we expect them and allow the test to fail if they exist in a different path and for some reason cause issues.

But will it really fail in that case? The reason we have this check in the first place is because the test won't automatically fail with wrong artifacts. It will happily continue even if they were built by a different version of solc. We force it to fail so that we can notice when the test is not really running with bytecode built by the custom-built solc binary we provide.

If neither of these two dirs exists, we should still hard fail, because the artifacts are probably still somewhre in there, just in a different location and we need to add that new location to the function.

Comment (r0qs) [reply scripts/externalTests/common.sh]: > But will it really fail in that case? The reason we have this check in the first place is because the test won't automatically fail with wrong artifacts. It will happily continue even if they were built by a different version of solc. We force it to fail so that we can notice when the test is not really running with bytecode built by the custom-built solc binary we provide.
> 
> If neither of these two dirs exists, we should still hard fail, because the artifacts are probably still somewhre in there, just in a different location and we need to add that new location to the function.

Oh, I see, I thought that it would fail if it used a wrong artifact. So yeah, if neither location is found we should thrown an error indeed. I will update the script.

Comment (cameel) [reply scripts/externalTests/common.sh]: Great. After you add back this safeguard, the PR will be fine to merge.

**Review by cameel — APPROVED** [review]

File: scripts/externalTests/common.sh (hunk 3 above)
Comment (r0qs) [reply scripts/externalTests/common.sh]: Done, I just added this line: `(( ${#artifact_dir[@]} != 0 )) || assertFail`
